R/parse_multiverse.R

```diff
@@ -74,30 +92,83 @@ get_parameter_conditions <- function(.expr) {
     # Base cases
     constant = , # falls through; the next element is evaluated
     symbol = list(parameters = list(), conditions = list()),
-    
+
     # Recursive cases
     call = {
       child_parameter_conditions <- map(.expr, get_parameter_conditions) %>%
         reduce(combine_parameter_conditions)
-      
+
       if (is_call(.expr, "branch")) {
           get_branch_parameter_conditions(.expr) %>%
-              combine_parameter_conditions(child_parameter_conditions)
+            combine_parameter_conditions(child_parameter_conditions)
+      } else if (is_call(.expr, "branch_assert")) {
+          get_branch_assert_condition(.expr) %>%
+            combine_parameter_conditions(child_parameter_conditions)
       } else {
           child_parameter_conditions
       }
     }
   )
 }
 
+# takes as input a `branch` call which contains a parameter name
+# and parameter processing options. Parameter option names is optional
+# returns as output a list(parameter = list(), condition = list())
+get_branch_parameter_conditions <- function(.branch_call) {
+  if (! is_symbol(.branch_call[[2]])) {
+    stop("parameter names should be symbols")
+  }
+  parameter_name <- .branch_call[[2]]
+  parameter_options <- map(.branch_call[-1:-2], ~ get_option_name(.x) )
+  parameter_conditions <- map(.branch_call[-1:-2], ~ get_condition(.x, parameter_name) )
+
+  parameter_options_list <- list(parameter_options)
+  names(parameter_options_list) <- as.character(parameter_name)
+
+  list( parameters = parameter_options_list, conditions = parameter_conditions )
+}
+
+get_branch_assert_condition <- function(.x) {
+  list(parameters = list(), conditions = list( expr((!!f_rhs(.x))) ))
+}
+
+get_condition <- function(.x, name) {
+  .antecedent = get_option_name(.x)
+  if (is_call(.x, "~")) {
+    .consequent = c(
+      get_implies_consequent(f_lhs(.x)),
+      get_implies_consequent(f_rhs(.x))
+    )[[1]]
+  } else {
+    .consequent = get_implies_consequent(.x)
+  }
+  if ( !is.null(.consequent)) {
+    expr(( !!name != !!.antecedent | !!.consequent ))
+  }
+}
+
+get_implies_consequent <- function(.x) {
+  if( is_call(.x, "%when%") ) {
+    f_rhs(.x)
+  } else {
+    if (is_call(.x, "(") | is_call(.x, "{")) {
+      get_implies_consequent(f_rhs(.x))
+    } else {
+      if (grepl("%when%",  expr_text(.x), fixed = TRUE) ) {
```

Comment:
*Never* use string matching to parse through code. Code already has structure. This approach is brittle (what if someone includes the string `"%when%"` in their code for some reason? Or what if they define their own `%when%` function that they want to use within their code?). 

If you just do nothing here, they should still get an error when the code executes later, right? Because `%when%` won't be a function defined in their code anyway.

Reply:
yeah, it'd throw an error if we do nothing. I'll remove this.